### ISSUE TITLE
"extension field may not have an explicitly set JSON name" error without JSON name set

### ISSUE THREAD

**Issue by LongHairedHacker** [golang/protobuf#1115]
**What version of protobuf and what language are you using?**
Version: `v1.22.0`

**What did you do?**
We recently started to upgrade some of our projects from `v.1.3.3` to `v1.22.0`, since we wanted to streamline some of the more repetitive parts of codebase using the new `protoreflect` API. 
However some of the proto files that worked fine before (and still work fine for other languages then golang) suddenly failed to compile.

Since I can not share the actual proto file in question, I've made minimal example to reproduce the issue.

**magic.proto**:
```
syntax = "proto3";
import "google/protobuf/descriptor.proto";


package MagicReproducer;


enum Magic
{
    unicorn = 0;
    rainbow = 1;
}

message MagicOption {
    Magic kind = 1;
}

extend google.protobuf.FieldOptions {
    MagicOption MO = 50000;
}
```

As a side note: I'm are aware that our naming does not exactly match the style guide
and we might start improving it in the future, but for now we are stuck with the current proto files.

I tried compiling the proto file using the following protoc command line:
```
$ protoc --go_out=plugins=grpc,Mgoogle/protobuf/descriptor.proto=google.golang.org/protobuf/protoc-gen-go/descriptor:./magic/ magic.proto 
```

**What did you expect to see?**
A magic.pb.go file that has been generated by protoc.

**What did you see instead?**
The following error message:

```
protoc-gen-go: invalid FileDescriptorProto "magic.proto": proto: extension field "MagicReproducer.MO" may not have an explicitly set JSON name: "mO"
```

The error message does not make much sense, as the JSON name is clearly not set explictly in the proto file.
The issue seems to be caused by the all uppercase name of the option.

As far as I can tell from reading the relevant source code in [reflect/protodesc/desc_validate.go](https://github.com/protocolbuffers/protobuf-go/blob/master/reflect/protodesc/desc_validate.go#L241) the cause of the problem ist that `GetJsonName()` returns `xXXXXX` for an all uppercase name, whereas `JSONCamelCase(string(x.Name()))` should return `XXXXXX` (if I'm not mistaken).

**Anything else we should know about your project / environment?**
```
$ go version
go version go1.14 linux/amd64
$ protoc --version
libprotoc 3.0.0
```

**Comment by dsnet** [comment]
`3.0.0` is a pretty old version of `protoc` (4 years old). Can you try something more recent? They're up to `3.11.0` and `3.12.0` is coming soon.

**Comment by dsnet** [comment]
I verified that this is a bug in `v3.0.0`, but it is fixed in `v3.1.0` of `protoc`.
